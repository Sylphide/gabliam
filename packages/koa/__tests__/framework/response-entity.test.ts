import {
  Controller,
  RestController,
  Get,
  ResponseEntity,
  ok,
  accepted,
  badRequest,
  noContent,
  notFound
} from '../../src';
import { KoaPluginTest } from '../koa-plugin-test';
import * as HttpStatus from 'http-status-codes';

let appTest: KoaPluginTest;

beforeEach(async () => {
  appTest = new KoaPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Response entity Tests:', () => {
  [Controller, RestController].forEach(decorator => {
    describe(`decorator ${decorator.name}`, () => {
      test('should work for responseEntitie controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          getTest() {
            return new ResponseEntity({ get: 'GET' });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test('should work for responseEntitie and custom Header controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          getTest() {
            const resp = new ResponseEntity({ get: 'GET' });
            resp.addHeader('X-Test', 'tests');
            return resp;
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test('should work for async responseEntitie controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            // console.log('ici');
            return new Promise(resolve => {
              setTimeout(resolve, 100, new ResponseEntity({ get: 'GET' }));
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.OK);

        expect(response).toMatchSnapshot();
      });

      test('responseEntitie ok()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return ok();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie ok({get: 'get'})`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return ok({ get: 'get' });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie accepted()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return accepted();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.ACCEPTED);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie badRequest()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return badRequest();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.BAD_REQUEST);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie noContent()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return noContent();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.NO_CONTENT);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie notFound()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return notFound();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(HttpStatus.NOT_FOUND);
        expect(response).toMatchSnapshot();
      });
    });
  });
});
