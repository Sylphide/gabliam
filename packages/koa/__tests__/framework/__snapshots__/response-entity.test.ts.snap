// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie accepted() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 202,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie badRequest() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 400,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie noContent() 1`] = `
Object {
  "header": Object {
    "connection": "close",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 204,
  "text": "",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie notFound() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie ok() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "2",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"\\"",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator responseEntitie ok({get: 'get'}) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"get\\"}",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator should work for async responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator should work for responseEntitie and custom Header controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "x-test": "tests",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator createControllerDecorator should work for responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie accepted() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 202,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie badRequest() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 400,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie noContent() 1`] = `
Object {
  "header": Object {
    "connection": "close",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 204,
  "text": "",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie notFound() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie ok() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "2",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"\\"",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator responseEntitie ok({get: 'get'}) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"get\\"}",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator should work for async responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator should work for responseEntitie and custom Header controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "x-test": "tests",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator createRestControllerDecorator should work for responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;
